### PR TITLE
Split up code and take some steps to reduce bundle size

### LINKED ISSUE #1500: combine javascript files that contain duplicate libraries
These are the current sizes of the JS files being served:

```
-rw------- 1 u46725 dyno  255221 Oct 21 15:43 course_list.js
-rw------- 1 u46725 dyno 2370309 Oct 21 15:43 dashboard.js
-rw------- 1 u46725 dyno  174782 Oct 21 15:43 faculty_carousel.js
-rw------- 1 u46725 dyno  431217 Oct 21 15:43 financial_aid.js
-rw------- 1 u46725 dyno   67623 Oct 21 15:43 public.js
-rw------- 1 u46725 dyno   24321 Oct 21 15:43 sentry_client.js
-rw------- 1 u46725 dyno  785638 Oct 21 15:43 signup_dialog.js
-rw------- 1 u46725 dyno  674183 Oct 21 15:43 style.js
-rw------- 1 u46725 dyno  229752 Oct 21 15:43 style_public.js
-rw------- 1 u46725 dyno    1197 Oct 21 15:43 zendesk_widget.js
```

### PR DESCRIPTION
**Pull request by aliceriot**
#### What are the relevant tickets?

closes #1500 

#### What's this PR do?

This makes some changes to our webpack config, to reduce the bundle size and make things (hopefully!) a little more efficient.

1. All JS code for the home page and the program pages are now combined into one script (`entry/public.js`).
2. I turned on several webpack plugins we probably should have been using, including one to do deduping and one for optimizing the order of chunks.
3. Modules that are common to all chunks are pulled out into a separate common module, which is required on all pages.

Here's what this gets us!

before (on `master`):

before_lol

after!

after_wow

#### How should this be manually tested?

Basically just make sure everything works and there are no style or JS related regressions anywhere. I think it would be good to test this in both development and in an environment where the JS has been built for production (like CI).

### CODE REVIEW

File: static/js/entry/public.js [hunk 1]
```diff
@@ -1,13 +1,69 @@
...
+    </MuiThemeProvider>,
+    courseListEl
+  );
+}
+
+// Program Page carousel div
+const carouselDiv = document.querySelector('#faculty-carousel');
+
+if ( carouselDiv ) {
+  if (carouselDiv !== null) {
```
Comment (noisecapella) [inline static/js/entry/public.js]: Can you remove one check or the other?

Comment (aliceriot) [inline static/js/entry/public.js]: oh haha copy-paste strikes again

File: webpack.config.prod.js [hunk 2]
```diff
@@ -20,7 +20,13 @@ module.exports = {
       compress: {
         warnings: false
       }
-    })
+    }),
+    new webpack.optimize.OccurrenceOrderPlugin(),
```
Comment (noisecapella) [inline webpack.config.prod.js]: Any reason these aren't included in `dev`?

Comment (aliceriot) [inline webpack.config.prod.js]: Well, mainly we don't care about file size in `dev`. I think the occurrence order one at least mangles the names of functions as well - although I'm not sure about that.
